### DISCUSSION TITLE
Super-droplet multiplicity after coalescence

### DISCUSSION THREAD

**Discussion by jbarr444** [open-atmos/PySDM#1380] [General]
Hi @slayoo,
Quick question: what happens to super-droplet multiplicities when they collide?
Thanks,
Jason

**Comment by slayoo** [comment]
Hi Jason,

Following step (5) in section 5.1.3 in [Shima et al. 2009](https://doi.org/10.1002/qj.441) (with multiplicities denoted by ξ):
![image](https://github.com/user-attachments/assets/64d6ffe8-5d61-4e2d-92bd-f6e09eb29be8)

which is reflected here for CPU:
https://github.com/open-atmos/PySDM/blob/cd9d7309b01e4138495455772e626b68bb8b6ecc/PySDM/backends/impl_numba/methods/collisions_methods.py#L49-L59

and here for GPU backend:
https://github.com/open-atmos/PySDM/blob/cd9d7309b01e4138495455772e626b68bb8b6ecc/PySDM/backends/impl_thrust_rtc/methods/collisions_methods.py#L42-L57

HTH,
S.

**Reply by jbarr444** [reply]
Very helpful, thanks Sylwester. I'm still a little confused though because this is for a constant super-droplet number, but I think in PySDM the number of super-droplets can decrease quite substantially, right?

**Reply by slayoo** [reply]
Please elaborate. I'd say that neither in the SDM paper, nor in PySDM, there is any assumption on constant number of super droplets.

**Reply by jbarr444** [reply]
Reading more thoroughly, the paper does say that super droplets can disappear when their multiplicity is 1. My bad. I'm starting to understand.

I am just interested in defining droplet ID, and hope that the 3 accessible attributes (position, radius, and multiplicity) are enough to attempt it roughly. Is this something that is being worked on by you or colleagues?

**Reply by slayoo** [reply]
As we've touched upon in #1360, the key to the `attributes` is the droplet ID, but droplet identity is lost upon collisions/breakup.
HTH,
S.

**Reply by jbarr444** [reply]
Do droplet indices shuffle around over time even if they don't collide, because droplets below them are  colliding and being removed from the system?

**Reply by slayoo** [reply]
The indices are not modified during the simulation.

**Reply by jbarr444** [reply]
I'm sorry, but I don't understand. It looks to me like the particles are always sorted by height bin, as the 'cell origin' attribute is always sorted. I can't find any evidence of a continuous particle path.
![image](https://github.com/user-attachments/assets/c5994c63-d947-4123-b92d-e918b631b0dc)
See how everything shifts abruptly at time index 100? That's exactly when my updraft ends, and I assume all super droplets in the reservoir are deleted. How could I trace a non-colliding particle through that break?
Thanks as always,
Jason

**Reply by slayoo** [reply]
@jbarr444, we'd need access to the source code generating the simulation and plot to comment - please, e.g. make a gist?

**Reply by jbarr444** [reply]
Sure, here's a link: https://gist.github.com/jbarr444/6c079927aa876b1ad42a4e679e4f014c
Note that I have edited the settings file to change the environmental conditions, but it shouldn't be important here.
Thanks!

**Reply by slayoo** [reply]
I've just had a quick look today, but it seems to me that this might help: https://github.com/open-atmos/PySDM/pull/1385 ?

**Reply by jbarr444** [reply]
I'm not sure what this page is saying. Does it mean this is a bug?

**Reply by slayoo** [reply]
Sorry for brevity. That PR introduces a change in the Shipway & Hill example in PySDM-examples package, which makes the stored attributes appear as-is in the output. Thus, if you apply the one-liner patch to PySDM-examples used for your case, the output should feature stable super droplet indices. It's hard to call it a bug or feature - one user would prefer the invalidated droplets to be hidden in the output, another would prefer preserving drop keys - both options have their rationale. I assume the best would be to make it an option easily controllable from user scrip level - will try to improve that PR in this direction. Please let us know if it that helps in your case?

**Reply by jbarr444** [reply]
Hi again @slayoo,
Following the equations you provided at the top of the discussion, I am attempting to detect collision partners by:
1) look for a particle whose multiplicity decreases, call the change in multiplicity dm
2) look for a particle with the same initial multiplicity and same subsequent volume; if none then
3) look for another particle whose multiplicity m is such that dm % m == 0
But I find quite often that there are no such candidate particles. How is this possible? Am I missing a case?
Thanks,
Jason

**Reply by slayoo** [reply]
@jbarr444, I wonder if the SDM adaptivity scheme that PySDM enables by default could be the cause here. Please try setting `coalescence_adaptive = False` in the S&H settings class.

**Reply by jbarr444** [reply]
This didn't work unfortunately.

**Reply by slayoo** [reply]
OK, here's Fig. 1b from https://doi.org/10.1007/978-3-030-77964-1_2:
![image](https://github.com/user-attachments/assets/6bdc20e3-3005-46fb-9838-445b86cdacd0)

Let's try to "thought-debug" this case:
1. the right-hand side particle has its multiplicity decreased from 9 to 1
2. the left-hand particle has its multiplicity unchanged, but the volume is bigger after collision - so not the case
3. dm=8, dm%2==0, so the left-hand one would be picked.

Is the above correct?

**Reply by jbarr444** [reply]
Yep, that's the idea!

**Reply by jbarr444** [reply]
What should be our next step in this thought-debug?
Hope not to bother @slayoo

**Reply by slayoo** [reply]
Sorry for the delay. Probably the most effective would be to post a gist in which an assert fails on what is expected to pass - then we can try to print-debug locally. Thanks

**Reply by jbarr444** [reply]
Here's a simplified version of my code: https://gist.github.com/jbarr444/76b4bdaafdcd82dc923d1f2f48286928
I'm not sure how to get my data to you. Perhaps the issue would arise in your own generated data?

**Reply by slayoo** [reply]
I'd say, we will be much better off creating some SD labeling logic within the backend methods, so that not only there would be no need for such "reverse-engineering", but also the results will be independent of the output frequency or adaptive time-stepping. We can introduce "counter" attributes that will track how many collisions, breakups and bounces happen to a super-droplet. This will enable us to use such attributes to compute the event-rate products (but only if needed, addressing #1149), and for your use case will provide flagging of particles that took part of collisions. Does it sound like a possible solution? Is knowing which SD took part in collision/breakup.bounce enough for your use case?

**Reply by jbarr444** [reply]
You’re absolutely right, labelling in the backend would be a much more robust way to accomplish this. If you would do that that would certainly solve the problem.

**Reply by slayoo** [reply]
@jbarr444, please check out this PR: https://github.com/open-atmos/PySDM/pull/1394

It introduces a new `FlagAttribute` logic with one sample implementation: `FlagCoalescence` 

There is a unit test depicting how to use it: https://github.com/slayoo/PySDM/blob/collision_flagging/tests/unit_tests/attributes/test_flags.py

In short:
```python
builder.request_attribute("flag coalescence")
```
and then one can access the flags with:
```python
particulator.attributes["flag coalescence"].to_ndarray()
```

If you have installed PySDM with `pip install -e` from a cloned repo, you can fetch the PR branch and do a checkout so that the new PR code would be imported. In case you are using PySDM not from a cloned repo, you can try out the new chages with `pip install git+https://github.com/slayoo/PySDM.git@collision_flagging`

**Reply by jbarr444** [reply]
Hi, thanks so much for doing this. I'm having some trouble installing this using 
pip install git+https://github.com/slayoo/PySDM.git@collision_flagging. 
The error, during "getting requirements to build the wheel", is 
"AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?"
... I don't suppose you have any idea what this is about? I'm using python 3.12.4 with anaconda.

**Reply by slayoo** [reply]
@jbarr444, does the usual `pip install PySDM` work in this environment?

**Reply by jbarr444** [reply]
Yes it does.

**Reply by slayoo** [reply]
There's nothing on that branch that would change version requirements or the installation procedure. Intriguing.
Please try doing a local install with:
```
git clone https://github.com/slayoo/PySDM.git
cd PySDM
git checkout collision_flagging
pip install -e .
```

**Reply by jbarr444** [reply]
This results in the same error unfortunately, but this time it was during 'Checking if build backend supports build_editable'.

**Reply by slayoo** [reply]
and if you do:
```
git checkout main
pip install -e .
```

**Reply by jbarr444** [reply]
Same error again.

**Reply by slayoo** [reply]
hmm... that confirms that it has nothing to do with the changes in the PR in question... 
This seems to be a report of a similar issue: https://github.com/pypa/pip/issues/12179
Does it also happen upon NumPy installation (as a dependency of PySDM) in your case?
Anyhow, please post the logs, it will be much easier to help

**Reply by jbarr444** [reply]
Well pip install numpy works, as does pip install PySDM. Perhaps it's an issue with git. Anyway, the full error output is:

C:\Users\jbpan\PySDM>pip install -e .
Processing c:\users\jbpan\pysdm
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [33 lines of output]
      Traceback (most recent call last):
        File "C:\Users\jbpan\anaconda3\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "C:\Users\jbpan\anaconda3\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\jbpan\anaconda3\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 112, in get_requires_for_build_wheel
          backend = _build_backend()
                    ^^^^^^^^^^^^^^^^
        File "C:\Users\jbpan\anaconda3\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 77, in _build_backend
          obj = import_module(mod_path)
                ^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\jbpan\anaconda3\Lib\importlib\__init__.py", line 90, in import_module
          return _bootstrap._gcd_import(name[level:], package, level)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1310, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
        File "<frozen importlib._bootstrap_external>", line 995, in exec_module
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "C:\Users\jbpan\AppData\Local\Temp\pip-build-env-mrw95gc6\overlay\Lib\site-packages\setuptools\__init__.py", line 16, in <module>
          import setuptools.version
        File "C:\Users\jbpan\AppData\Local\Temp\pip-build-env-mrw95gc6\overlay\Lib\site-packages\setuptools\version.py", line 1, in <module>
          import pkg_resources
        File "C:\Users\jbpan\AppData\Local\Temp\pip-build-env-mrw95gc6\overlay\Lib\site-packages\pkg_resources\__init__.py", line 2158, in <module>
          register_finder(pkgutil.ImpImporter, find_on_path)
                          ^^^^^^^^^^^^^^^^^^^
      AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

**Reply by slayoo** [reply]
Googling more, there seem to be a recurrent suggestion to do ``pip install setuptools`` (and in fact we do it on CI for Python 3.12 in PySDM) - could you please try if that helps?

**Reply by jbarr444** [reply]
Yes, I read the same a few hours ago and did that before all of your help today.

I just made a new environment of python 3.11 and it installed successfully. It will take a bit for me to set everything back up but hopefully this will work for now.

**Reply by slayoo** [reply]
Great that you found a workaround, but apparently we need to improve CI coverage of Python 3.12 testing!

**Reply by jbarr444** [reply]
So your test ran without error, but this python 3.11 thing is screwing up my main code. Vanilla Shipway & Hill runs but my modified version doesn't for some reason. I use Formulae.saturation_vapor_pressure.pvs_Celsius and that seems to have broken, among other things. I think I'll keep trying to tamper with 3.12 to get it to install I guess. Anyways just letting you know, I'll try to fix it myself.

**Reply by slayoo** [reply]
since PySDM 2.77 (https://github.com/open-atmos/PySDM/releases/tag/v2.77), the saturation vapour pressure functions are now all-SI unit - so no more Celsius. It should be enough to switch from `pvs_Celsius(T)` to `pvs_water(T+T0)` in your code. This issue is not related with the version of Python. HTH, S.

**Reply by jbarr444** [reply]
Oh, wonderful, that fixes it completely. That's strange though because it doesn't upgrade to 2.77 when I pip install PySDM on python 3.12, but it is 2.77 on python 3.11. Thanks anyways.

**Reply by slayoo** [reply]
Here's the reason - there was a timeout in the PyPI upload job for 2.77, so it has not yet reached PyPI. I've just trigerred a rerun, so it should be fixed in an hour: https://github.com/open-atmos/PySDM/actions/runs/10916574488
Thanks for notifying about it!

**Reply by slayoo** [reply]
For the record, the upload rerun on Fri went OK and v2.77 is there on PyPI https://pypi.org/project/PySDM/2.77/

**Reply by jbarr444** [reply]
Updated, thanks!

**Comment by jbarr444** [comment]
@slayoo That works !! Thank you :)
